pkg/tsdb/interval/interval.go

```diff
@@ -54,16 +55,37 @@ func (i *Interval) Milliseconds() int64 {
 	return i.Value.Nanoseconds() / int64(time.Millisecond)
 }
 
-func (ic *intervalCalculator) Calculate(timerange plugins.DataTimeRange, minInterval time.Duration) Interval {
+func (ic *intervalCalculator) Calculate(timerange plugins.DataTimeRange, interval time.Duration, intervalMode string) Interval {
 	to := timerange.MustGetTo().UnixNano()
 	from := timerange.MustGetFrom().UnixNano()
-	interval := time.Duration((to - from) / defaultRes)
+	calculatedInterval := time.Duration((to - from) / defaultRes)
 
-	if interval < minInterval {
-		return Interval{Text: FormatDuration(minInterval), Value: minInterval}
+	switch intervalMode {
+	case "min":
+		if calculatedInterval < interval {
+			return Interval{Text: FormatDuration(interval), Value: interval}
+		}
+	case "max":
+		if calculatedInterval > interval {
+			return Interval{Text: FormatDuration(interval), Value: interval}
+		}
+	case "exact":
+		return Interval{Text: FormatDuration(interval), Value: interval}
+
+	default:
+		panic(fmt.Sprintf("Unrecognized intervalMode %v", intervalMode))
```

Comment:
As I mentioned I would return the error or log it

Reply:
Updated in https://github.com/grafana/grafana/pull/36796/commits/c42ebe7d3f2b71d385e26fcac48de5699df70a03